// analyze-desumasu-dearu/lib/analyze.js
// LICENSE : MIT
"use strict";
// This RegExp are based of https://github.com/recruit-tech/redpen/blob/master/redpen-core/src/main/java/cc/redpen/validator/sentence/JapaneseStyleValidator.java

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.analyzeDesumasu = analyzeDesumasu;
exports.analyzeDearu = analyzeDearu;
var DEARU_PATTERN = /のだが|ないかと|であるから/g;
var DEARU_END_PATTERN = /(だ|である|った|ではない｜ないか|しろ|しなさい|いただきたい|いただく|ならない|あろう|られる)。/;

var DESUMASU_PATTERN = /でしたが|でしたので|ですので|ですが/g;
var DESUMASU_END_PATTERN = /(です|ます|ました|ません|ですね|でしょうか|ください|ませ)。/;
/**
 *
 * @param text
 * @param reg
 * @returns {{value:string, columnIndex: number, lineNumber:number}}[]
 */
function countMatchContent(text, reg) {
    var matches = [];
    var tmpMatch = undefined;
    while ((tmpMatch = reg.exec(text)) != null) {
        matches.push({
            value: tmpMatch[0],
            lineNumber: 1,
            columnIndex: reg.lastIndex - tmpMatch[0].length
        });
    }
    return matches;
}
/**
 *
 * @param text
 * @param reg
 * @returns {{value:string, columnIndex: number, lineNumber:number}}[]
 */
function countMatchContentEnd(text, reg) {
    var lines = text.split(/\r\n|\r|\n|\u2028|\u2029/g);
    var matches = [];
    lines.forEach(function (line, index) {
        var match = line.match(reg);
        if (!match) {
            return;
        }
        // adjust line number
        matches.push({
            value: match[0],
            lineNumber: 1 + index,
            columnIndex: match.index
        });
    });
    return matches;
}
/**
 * `text` の敬体(ですます調)について解析します
 * @param {string} text
 * @param {object} options
 * @param {boolean} options.analyzeConjunction 接続詞を解析するかどうか default: true
 * @returns {{value:string, columnIndex: number, lineNumber:number}}[]
 */

function analyzeDesumasu(text) {
    var options = arguments.length <= 1 || arguments[1] === undefined ? { analyzeConjunction: true } : arguments[1];
    var analyzeConjunction = options.analyzeConjunction;

    if (!analyzeConjunction) {
        return countMatchContentEnd(text, DESUMASU_END_PATTERN);
    }
    var retDesumasu = countMatchContent(text, DESUMASU_PATTERN);
    var retDesumasuEnd = countMatchContentEnd(text, DESUMASU_END_PATTERN);
    return retDesumasu.concat(retDesumasuEnd);
}
/**
 * `text` の常体(である調)について解析します
 * @param {string} text
 * @param {object} options
 * @param {boolean} options.analyzeConjunction 接続詞を解析するかどうか default: true
 * @returns {{value:string, columnIndex: number, lineNumber:number}}[]
 */
function analyzeDearu(text) {
    var options = arguments.length <= 1 || arguments[1] === undefined ? { analyzeConjunction: true } : arguments[1];
    var analyzeConjunction = options.analyzeConjunction;

    if (!analyzeConjunction) {
        return countMatchContentEnd(text, DEARU_END_PATTERN);
    }
    var retDearu = countMatchContent(text, DEARU_PATTERN);
    var retDearuEnd = countMatchContentEnd(text, DEARU_END_PATTERN);
    return retDearu.concat(retDearuEnd);
}
