// @kvs/storage/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.hasItem = exports.createIterator = exports.clearItem = exports.getItem = exports.setItem = exports.deleteItem = exports.kvsStorage = void 0;
var storage_1 = require("./storage");
Object.defineProperty(exports, "kvsStorage", { enumerable: true, get: function () { return storage_1.kvsStorage; } });
Object.defineProperty(exports, "deleteItem", { enumerable: true, get: function () { return storage_1.deleteItem; } });
Object.defineProperty(exports, "setItem", { enumerable: true, get: function () { return storage_1.setItem; } });
Object.defineProperty(exports, "getItem", { enumerable: true, get: function () { return storage_1.getItem; } });
Object.defineProperty(exports, "clearItem", { enumerable: true, get: function () { return storage_1.clearItem; } });
Object.defineProperty(exports, "createIterator", { enumerable: true, get: function () { return storage_1.createIterator; } });
Object.defineProperty(exports, "hasItem", { enumerable: true, get: function () { return storage_1.hasItem; } });
