// @textlint/kernel/module/src/descriptor/TextlintFilterRuleDescriptors.js
// LICENSE : MIT
"use strict";
import { filterDuplicateDescriptor } from "./DescriptorUtil";
/**
 * The collection class of TextlintFilterRuleDescriptor
 */
export class TextlintFilterRuleDescriptors {
    constructor(ruleDescriptorList = []) {
        this.ruleDescriptorList = ruleDescriptorList;
    }
    /**
     * Convert this to TextlintKernel rules format
     * @returns {Array}
     */
    toKernelFilterRulesFormat() {
        return this.withoutDuplicated().descriptors.map((descriptor) => {
            return descriptor.toKernel();
        });
    }
    /**
     * Return enabled descriptors
     */
    get descriptors() {
        return this.ruleDescriptorList.filter((descriptor) => {
            return descriptor.enabled;
        });
    }
    /**
     * Return all descriptors that include disabled descriptors
     */
    get allDescriptors() {
        return this.ruleDescriptorList;
    }
    /**
     * filter duplicated descriptors
     */
    withoutDuplicated() {
        // remove last duplicated item
        const newDescriptorList = filterDuplicateDescriptor(this.ruleDescriptorList);
        return new TextlintFilterRuleDescriptors(newDescriptorList);
    }
}
