// nlcst-types/lib/type-guard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isTextNode = exports.isSource = exports.isWhiteSpace = exports.isPunctuation = exports.isSymbol = exports.isText = exports.isWord = exports.isSentence = exports.isParagraph = exports.isRoot = void 0;
var isRoot = function (v) {
    return v && v.type === "RootNode";
};
exports.isRoot = isRoot;
var isParagraph = function (v) {
    return v && v.type === "ParagraphNode";
};
exports.isParagraph = isParagraph;
var isSentence = function (v) {
    return v && v.type === "SentenceNode";
};
exports.isSentence = isSentence;
var isWord = function (v) {
    return v && v.type === "WordNode";
};
exports.isWord = isWord;
var isText = function (v) {
    return v && v.type === "TextNode";
};
exports.isText = isText;
var isSymbol = function (v) {
    return v && v.type === "SymbolNode";
};
exports.isSymbol = isSymbol;
var isPunctuation = function (v) {
    return v && v.type === "PunctuationNode";
};
exports.isPunctuation = isPunctuation;
var isWhiteSpace = function (v) {
    return v && v.type === "WhiteSpaceNode";
};
exports.isWhiteSpace = isWhiteSpace;
var isSource = function (v) {
    return v && v.type === "SourceNode";
};
exports.isSource = isSource;
var isTextNode = function (v) {
    return v && v.type === "TextNode";
};
exports.isTextNode = isTextNode;
