// @aws-sdk/client-ec2/dist-cjs/commands/DescribeNetworkInsightsAccessScopesCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DescribeNetworkInsightsAccessScopesCommand = void 0;
const middleware_serde_1 = require("@aws-sdk/middleware-serde");
const smithy_client_1 = require("@aws-sdk/smithy-client");
const models_3_1 = require("../models/models_3");
const Aws_ec2_1 = require("../protocols/Aws_ec2");
class DescribeNetworkInsightsAccessScopesCommand extends smithy_client_1.Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use((0, middleware_serde_1.getSerdePlugin)(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "EC2Client";
        const commandName = "DescribeNetworkInsightsAccessScopesCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: models_3_1.DescribeNetworkInsightsAccessScopesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: models_3_1.DescribeNetworkInsightsAccessScopesResult.filterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return (0, Aws_ec2_1.serializeAws_ec2DescribeNetworkInsightsAccessScopesCommand)(input, context);
    }
    deserialize(output, context) {
        return (0, Aws_ec2_1.deserializeAws_ec2DescribeNetworkInsightsAccessScopesCommand)(output, context);
    }
}
exports.DescribeNetworkInsightsAccessScopesCommand = DescribeNetworkInsightsAccessScopesCommand;
