// @aws-sdk/client-ec2/dist-cjs/commands/GetAssociatedIpv6PoolCidrsCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetAssociatedIpv6PoolCidrsCommand = void 0;
const middleware_serde_1 = require("@aws-sdk/middleware-serde");
const smithy_client_1 = require("@aws-sdk/smithy-client");
const models_4_1 = require("../models/models_4");
const Aws_ec2_1 = require("../protocols/Aws_ec2");
class GetAssociatedIpv6PoolCidrsCommand extends smithy_client_1.Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use((0, middleware_serde_1.getSerdePlugin)(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "EC2Client";
        const commandName = "GetAssociatedIpv6PoolCidrsCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: models_4_1.GetAssociatedIpv6PoolCidrsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: models_4_1.GetAssociatedIpv6PoolCidrsResult.filterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return (0, Aws_ec2_1.serializeAws_ec2GetAssociatedIpv6PoolCidrsCommand)(input, context);
    }
    deserialize(output, context) {
        return (0, Aws_ec2_1.deserializeAws_ec2GetAssociatedIpv6PoolCidrsCommand)(output, context);
    }
}
exports.GetAssociatedIpv6PoolCidrsCommand = GetAssociatedIpv6PoolCidrsCommand;
