// @aws-sdk/client-ec2/dist-cjs/commands/ModifySecurityGroupRulesCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModifySecurityGroupRulesCommand = void 0;
const middleware_serde_1 = require("@aws-sdk/middleware-serde");
const smithy_client_1 = require("@aws-sdk/smithy-client");
const models_5_1 = require("../models/models_5");
const Aws_ec2_1 = require("../protocols/Aws_ec2");
class ModifySecurityGroupRulesCommand extends smithy_client_1.Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use((0, middleware_serde_1.getSerdePlugin)(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "EC2Client";
        const commandName = "ModifySecurityGroupRulesCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: models_5_1.ModifySecurityGroupRulesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: models_5_1.ModifySecurityGroupRulesResult.filterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return (0, Aws_ec2_1.serializeAws_ec2ModifySecurityGroupRulesCommand)(input, context);
    }
    deserialize(output, context) {
        return (0, Aws_ec2_1.deserializeAws_ec2ModifySecurityGroupRulesCommand)(output, context);
    }
}
exports.ModifySecurityGroupRulesCommand = ModifySecurityGroupRulesCommand;
