// @aws-sdk/client-ec2/dist-cjs/commands/UnassignIpv6AddressesCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnassignIpv6AddressesCommand = void 0;
const middleware_serde_1 = require("@aws-sdk/middleware-serde");
const smithy_client_1 = require("@aws-sdk/smithy-client");
const models_6_1 = require("../models/models_6");
const Aws_ec2_1 = require("../protocols/Aws_ec2");
class UnassignIpv6AddressesCommand extends smithy_client_1.Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use((0, middleware_serde_1.getSerdePlugin)(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "EC2Client";
        const commandName = "UnassignIpv6AddressesCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: models_6_1.UnassignIpv6AddressesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: models_6_1.UnassignIpv6AddressesResult.filterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return (0, Aws_ec2_1.serializeAws_ec2UnassignIpv6AddressesCommand)(input, context);
    }
    deserialize(output, context) {
        return (0, Aws_ec2_1.deserializeAws_ec2UnassignIpv6AddressesCommand)(output, context);
    }
}
exports.UnassignIpv6AddressesCommand = UnassignIpv6AddressesCommand;
