// @aws-sdk/client-ecr/dist-es/commands/BatchGetRepositoryScanningConfigurationCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { BatchGetRepositoryScanningConfigurationRequest, BatchGetRepositoryScanningConfigurationResponse, } from "../models/models_0";
import { deserializeAws_json1_1BatchGetRepositoryScanningConfigurationCommand, serializeAws_json1_1BatchGetRepositoryScanningConfigurationCommand, } from "../protocols/Aws_json1_1";
var BatchGetRepositoryScanningConfigurationCommand = (function (_super) {
    __extends(BatchGetRepositoryScanningConfigurationCommand, _super);
    function BatchGetRepositoryScanningConfigurationCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    BatchGetRepositoryScanningConfigurationCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ECRClient";
        var commandName = "BatchGetRepositoryScanningConfigurationCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: BatchGetRepositoryScanningConfigurationRequest.filterSensitiveLog,
            outputFilterSensitiveLog: BatchGetRepositoryScanningConfigurationResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    BatchGetRepositoryScanningConfigurationCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1BatchGetRepositoryScanningConfigurationCommand(input, context);
    };
    BatchGetRepositoryScanningConfigurationCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1BatchGetRepositoryScanningConfigurationCommand(output, context);
    };
    return BatchGetRepositoryScanningConfigurationCommand;
}($Command));
export { BatchGetRepositoryScanningConfigurationCommand };
