// @aws-sdk/client-ecr/dist-es/commands/DeleteRegistryPolicyCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteRegistryPolicyRequest, DeleteRegistryPolicyResponse } from "../models/models_0";
import { deserializeAws_json1_1DeleteRegistryPolicyCommand, serializeAws_json1_1DeleteRegistryPolicyCommand, } from "../protocols/Aws_json1_1";
var DeleteRegistryPolicyCommand = (function (_super) {
    __extends(DeleteRegistryPolicyCommand, _super);
    function DeleteRegistryPolicyCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteRegistryPolicyCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ECRClient";
        var commandName = "DeleteRegistryPolicyCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteRegistryPolicyRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DeleteRegistryPolicyResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteRegistryPolicyCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DeleteRegistryPolicyCommand(input, context);
    };
    DeleteRegistryPolicyCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DeleteRegistryPolicyCommand(output, context);
    };
    return DeleteRegistryPolicyCommand;
}($Command));
export { DeleteRegistryPolicyCommand };
