// @aws-sdk/client-ecr/dist-es/commands/DescribeImageScanFindingsCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeImageScanFindingsRequest, DescribeImageScanFindingsResponse } from "../models/models_0";
import { deserializeAws_json1_1DescribeImageScanFindingsCommand, serializeAws_json1_1DescribeImageScanFindingsCommand, } from "../protocols/Aws_json1_1";
var DescribeImageScanFindingsCommand = (function (_super) {
    __extends(DescribeImageScanFindingsCommand, _super);
    function DescribeImageScanFindingsCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeImageScanFindingsCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ECRClient";
        var commandName = "DescribeImageScanFindingsCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeImageScanFindingsRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DescribeImageScanFindingsResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeImageScanFindingsCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeImageScanFindingsCommand(input, context);
    };
    DescribeImageScanFindingsCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeImageScanFindingsCommand(output, context);
    };
    return DescribeImageScanFindingsCommand;
}($Command));
export { DescribeImageScanFindingsCommand };
