// @aws-sdk/client-ecr/dist-es/commands/DescribeRepositoriesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeRepositoriesRequest, DescribeRepositoriesResponse } from "../models/models_0";
import { deserializeAws_json1_1DescribeRepositoriesCommand, serializeAws_json1_1DescribeRepositoriesCommand, } from "../protocols/Aws_json1_1";
var DescribeRepositoriesCommand = (function (_super) {
    __extends(DescribeRepositoriesCommand, _super);
    function DescribeRepositoriesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeRepositoriesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ECRClient";
        var commandName = "DescribeRepositoriesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeRepositoriesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DescribeRepositoriesResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeRepositoriesCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeRepositoriesCommand(input, context);
    };
    DescribeRepositoriesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeRepositoriesCommand(output, context);
    };
    return DescribeRepositoriesCommand;
}($Command));
export { DescribeRepositoriesCommand };
