// @aws-sdk/client-ecr/dist-es/commands/DescribeRegistryCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeRegistryRequest, DescribeRegistryResponse } from "../models/models_0";
import { deserializeAws_json1_1DescribeRegistryCommand, serializeAws_json1_1DescribeRegistryCommand, } from "../protocols/Aws_json1_1";
var DescribeRegistryCommand = (function (_super) {
    __extends(DescribeRegistryCommand, _super);
    function DescribeRegistryCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeRegistryCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ECRClient";
        var commandName = "DescribeRegistryCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeRegistryRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DescribeRegistryResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeRegistryCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1DescribeRegistryCommand(input, context);
    };
    DescribeRegistryCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1DescribeRegistryCommand(output, context);
    };
    return DescribeRegistryCommand;
}($Command));
export { DescribeRegistryCommand };
