// @aws-sdk/client-ecr/dist-es/commands/PutLifecyclePolicyCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutLifecyclePolicyRequest, PutLifecyclePolicyResponse } from "../models/models_0";
import { deserializeAws_json1_1PutLifecyclePolicyCommand, serializeAws_json1_1PutLifecyclePolicyCommand, } from "../protocols/Aws_json1_1";
var PutLifecyclePolicyCommand = (function (_super) {
    __extends(PutLifecyclePolicyCommand, _super);
    function PutLifecyclePolicyCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutLifecyclePolicyCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ECRClient";
        var commandName = "PutLifecyclePolicyCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutLifecyclePolicyRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutLifecyclePolicyResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutLifecyclePolicyCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1PutLifecyclePolicyCommand(input, context);
    };
    PutLifecyclePolicyCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1PutLifecyclePolicyCommand(output, context);
    };
    return PutLifecyclePolicyCommand;
}($Command));
export { PutLifecyclePolicyCommand };
