// @aws-sdk/client-ecr/dist-es/commands/PutRegistryPolicyCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutRegistryPolicyRequest, PutRegistryPolicyResponse } from "../models/models_0";
import { deserializeAws_json1_1PutRegistryPolicyCommand, serializeAws_json1_1PutRegistryPolicyCommand, } from "../protocols/Aws_json1_1";
var PutRegistryPolicyCommand = (function (_super) {
    __extends(PutRegistryPolicyCommand, _super);
    function PutRegistryPolicyCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutRegistryPolicyCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ECRClient";
        var commandName = "PutRegistryPolicyCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutRegistryPolicyRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutRegistryPolicyResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutRegistryPolicyCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1PutRegistryPolicyCommand(input, context);
    };
    PutRegistryPolicyCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1PutRegistryPolicyCommand(output, context);
    };
    return PutRegistryPolicyCommand;
}($Command));
export { PutRegistryPolicyCommand };
