// @aws-sdk/client-ecr/dist-es/commands/PutReplicationConfigurationCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { PutReplicationConfigurationRequest, PutReplicationConfigurationResponse } from "../models/models_0";
import { deserializeAws_json1_1PutReplicationConfigurationCommand, serializeAws_json1_1PutReplicationConfigurationCommand, } from "../protocols/Aws_json1_1";
var PutReplicationConfigurationCommand = (function (_super) {
    __extends(PutReplicationConfigurationCommand, _super);
    function PutReplicationConfigurationCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    PutReplicationConfigurationCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ECRClient";
        var commandName = "PutReplicationConfigurationCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: PutReplicationConfigurationRequest.filterSensitiveLog,
            outputFilterSensitiveLog: PutReplicationConfigurationResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    PutReplicationConfigurationCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1PutReplicationConfigurationCommand(input, context);
    };
    PutReplicationConfigurationCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1PutReplicationConfigurationCommand(output, context);
    };
    return PutReplicationConfigurationCommand;
}($Command));
export { PutReplicationConfigurationCommand };
