// @aws-sdk/client-ecr/dist-es/commands/StartLifecyclePolicyPreviewCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { StartLifecyclePolicyPreviewRequest, StartLifecyclePolicyPreviewResponse } from "../models/models_0";
import { deserializeAws_json1_1StartLifecyclePolicyPreviewCommand, serializeAws_json1_1StartLifecyclePolicyPreviewCommand, } from "../protocols/Aws_json1_1";
var StartLifecyclePolicyPreviewCommand = (function (_super) {
    __extends(StartLifecyclePolicyPreviewCommand, _super);
    function StartLifecyclePolicyPreviewCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    StartLifecyclePolicyPreviewCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "ECRClient";
        var commandName = "StartLifecyclePolicyPreviewCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: StartLifecyclePolicyPreviewRequest.filterSensitiveLog,
            outputFilterSensitiveLog: StartLifecyclePolicyPreviewResponse.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    StartLifecyclePolicyPreviewCommand.prototype.serialize = function (input, context) {
        return serializeAws_json1_1StartLifecyclePolicyPreviewCommand(input, context);
    };
    StartLifecyclePolicyPreviewCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_json1_1StartLifecyclePolicyPreviewCommand(output, context);
    };
    return StartLifecyclePolicyPreviewCommand;
}($Command));
export { StartLifecyclePolicyPreviewCommand };
