// @aws-sdk/client-ec2/dist-es/commands/CopyFpgaImageCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { CopyFpgaImageRequest, CopyFpgaImageResult } from "../models/models_0";
import { deserializeAws_ec2CopyFpgaImageCommand, serializeAws_ec2CopyFpgaImageCommand } from "../protocols/Aws_ec2";
var CopyFpgaImageCommand = (function (_super) {
    __extends(CopyFpgaImageCommand, _super);
    function CopyFpgaImageCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    CopyFpgaImageCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "EC2Client";
        var commandName = "CopyFpgaImageCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: CopyFpgaImageRequest.filterSensitiveLog,
            outputFilterSensitiveLog: CopyFpgaImageResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    CopyFpgaImageCommand.prototype.serialize = function (input, context) {
        return serializeAws_ec2CopyFpgaImageCommand(input, context);
    };
    CopyFpgaImageCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_ec2CopyFpgaImageCommand(output, context);
    };
    return CopyFpgaImageCommand;
}($Command));
export { CopyFpgaImageCommand };
