// @aws-sdk/client-ec2/dist-es/commands/DeleteRouteCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteRouteRequest } from "../models/models_2";
import { deserializeAws_ec2DeleteRouteCommand, serializeAws_ec2DeleteRouteCommand } from "../protocols/Aws_ec2";
var DeleteRouteCommand = (function (_super) {
    __extends(DeleteRouteCommand, _super);
    function DeleteRouteCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteRouteCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "EC2Client";
        var commandName = "DeleteRouteCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteRouteRequest.filterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteRouteCommand.prototype.serialize = function (input, context) {
        return serializeAws_ec2DeleteRouteCommand(input, context);
    };
    DeleteRouteCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_ec2DeleteRouteCommand(output, context);
    };
    return DeleteRouteCommand;
}($Command));
export { DeleteRouteCommand };
