// @aws-sdk/client-ec2/dist-es/commands/DeleteVpcCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DeleteVpcRequest } from "../models/models_2";
import { deserializeAws_ec2DeleteVpcCommand, serializeAws_ec2DeleteVpcCommand } from "../protocols/Aws_ec2";
var DeleteVpcCommand = (function (_super) {
    __extends(DeleteVpcCommand, _super);
    function DeleteVpcCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DeleteVpcCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "EC2Client";
        var commandName = "DeleteVpcCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DeleteVpcRequest.filterSensitiveLog,
            outputFilterSensitiveLog: function (output) { return output; },
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DeleteVpcCommand.prototype.serialize = function (input, context) {
        return serializeAws_ec2DeleteVpcCommand(input, context);
    };
    DeleteVpcCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_ec2DeleteVpcCommand(output, context);
    };
    return DeleteVpcCommand;
}($Command));
export { DeleteVpcCommand };
