// @aws-sdk/client-ec2/dist-es/commands/DescribeFastLaunchImagesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DescribeFastLaunchImagesRequest, DescribeFastLaunchImagesResult } from "../models/models_3";
import { deserializeAws_ec2DescribeFastLaunchImagesCommand, serializeAws_ec2DescribeFastLaunchImagesCommand, } from "../protocols/Aws_ec2";
var DescribeFastLaunchImagesCommand = (function (_super) {
    __extends(DescribeFastLaunchImagesCommand, _super);
    function DescribeFastLaunchImagesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DescribeFastLaunchImagesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "EC2Client";
        var commandName = "DescribeFastLaunchImagesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DescribeFastLaunchImagesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DescribeFastLaunchImagesResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DescribeFastLaunchImagesCommand.prototype.serialize = function (input, context) {
        return serializeAws_ec2DescribeFastLaunchImagesCommand(input, context);
    };
    DescribeFastLaunchImagesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_ec2DescribeFastLaunchImagesCommand(output, context);
    };
    return DescribeFastLaunchImagesCommand;
}($Command));
export { DescribeFastLaunchImagesCommand };
