// @aws-sdk/client-ec2/dist-es/commands/DisableFastLaunchCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { DisableFastLaunchRequest, DisableFastLaunchResult } from "../models/models_4";
import { deserializeAws_ec2DisableFastLaunchCommand, serializeAws_ec2DisableFastLaunchCommand, } from "../protocols/Aws_ec2";
var DisableFastLaunchCommand = (function (_super) {
    __extends(DisableFastLaunchCommand, _super);
    function DisableFastLaunchCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    DisableFastLaunchCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "EC2Client";
        var commandName = "DisableFastLaunchCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: DisableFastLaunchRequest.filterSensitiveLog,
            outputFilterSensitiveLog: DisableFastLaunchResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    DisableFastLaunchCommand.prototype.serialize = function (input, context) {
        return serializeAws_ec2DisableFastLaunchCommand(input, context);
    };
    DisableFastLaunchCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_ec2DisableFastLaunchCommand(output, context);
    };
    return DisableFastLaunchCommand;
}($Command));
export { DisableFastLaunchCommand };
