// @aws-sdk/client-ec2/dist-es/commands/EnableFastLaunchCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { EnableFastLaunchRequest, EnableFastLaunchResult } from "../models/models_4";
import { deserializeAws_ec2EnableFastLaunchCommand, serializeAws_ec2EnableFastLaunchCommand, } from "../protocols/Aws_ec2";
var EnableFastLaunchCommand = (function (_super) {
    __extends(EnableFastLaunchCommand, _super);
    function EnableFastLaunchCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    EnableFastLaunchCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "EC2Client";
        var commandName = "EnableFastLaunchCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: EnableFastLaunchRequest.filterSensitiveLog,
            outputFilterSensitiveLog: EnableFastLaunchResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    EnableFastLaunchCommand.prototype.serialize = function (input, context) {
        return serializeAws_ec2EnableFastLaunchCommand(input, context);
    };
    EnableFastLaunchCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_ec2EnableFastLaunchCommand(output, context);
    };
    return EnableFastLaunchCommand;
}($Command));
export { EnableFastLaunchCommand };
