// @aws-sdk/client-ec2/dist-es/commands/ListImagesInRecycleBinCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ListImagesInRecycleBinRequest, ListImagesInRecycleBinResult } from "../models/models_5";
import { deserializeAws_ec2ListImagesInRecycleBinCommand, serializeAws_ec2ListImagesInRecycleBinCommand, } from "../protocols/Aws_ec2";
var ListImagesInRecycleBinCommand = (function (_super) {
    __extends(ListImagesInRecycleBinCommand, _super);
    function ListImagesInRecycleBinCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ListImagesInRecycleBinCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "EC2Client";
        var commandName = "ListImagesInRecycleBinCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ListImagesInRecycleBinRequest.filterSensitiveLog,
            outputFilterSensitiveLog: ListImagesInRecycleBinResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ListImagesInRecycleBinCommand.prototype.serialize = function (input, context) {
        return serializeAws_ec2ListImagesInRecycleBinCommand(input, context);
    };
    ListImagesInRecycleBinCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_ec2ListImagesInRecycleBinCommand(output, context);
    };
    return ListImagesInRecycleBinCommand;
}($Command));
export { ListImagesInRecycleBinCommand };
