// @aws-sdk/client-ec2/dist-es/commands/ModifyVpcEndpointServicePayerResponsibilityCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ModifyVpcEndpointServicePayerResponsibilityRequest, ModifyVpcEndpointServicePayerResponsibilityResult, } from "../models/models_5";
import { deserializeAws_ec2ModifyVpcEndpointServicePayerResponsibilityCommand, serializeAws_ec2ModifyVpcEndpointServicePayerResponsibilityCommand, } from "../protocols/Aws_ec2";
var ModifyVpcEndpointServicePayerResponsibilityCommand = (function (_super) {
    __extends(ModifyVpcEndpointServicePayerResponsibilityCommand, _super);
    function ModifyVpcEndpointServicePayerResponsibilityCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ModifyVpcEndpointServicePayerResponsibilityCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "EC2Client";
        var commandName = "ModifyVpcEndpointServicePayerResponsibilityCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ModifyVpcEndpointServicePayerResponsibilityRequest.filterSensitiveLog,
            outputFilterSensitiveLog: ModifyVpcEndpointServicePayerResponsibilityResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ModifyVpcEndpointServicePayerResponsibilityCommand.prototype.serialize = function (input, context) {
        return serializeAws_ec2ModifyVpcEndpointServicePayerResponsibilityCommand(input, context);
    };
    ModifyVpcEndpointServicePayerResponsibilityCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_ec2ModifyVpcEndpointServicePayerResponsibilityCommand(output, context);
    };
    return ModifyVpcEndpointServicePayerResponsibilityCommand;
}($Command));
export { ModifyVpcEndpointServicePayerResponsibilityCommand };
