// @aws-sdk/client-ec2/dist-es/commands/ModifyTrafficMirrorFilterNetworkServicesCommand.js
import { __extends } from "tslib";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { ModifyTrafficMirrorFilterNetworkServicesRequest, ModifyTrafficMirrorFilterNetworkServicesResult, } from "../models/models_5";
import { deserializeAws_ec2ModifyTrafficMirrorFilterNetworkServicesCommand, serializeAws_ec2ModifyTrafficMirrorFilterNetworkServicesCommand, } from "../protocols/Aws_ec2";
var ModifyTrafficMirrorFilterNetworkServicesCommand = (function (_super) {
    __extends(ModifyTrafficMirrorFilterNetworkServicesCommand, _super);
    function ModifyTrafficMirrorFilterNetworkServicesCommand(input) {
        var _this = _super.call(this) || this;
        _this.input = input;
        return _this;
    }
    ModifyTrafficMirrorFilterNetworkServicesCommand.prototype.resolveMiddleware = function (clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        var stack = clientStack.concat(this.middlewareStack);
        var logger = configuration.logger;
        var clientName = "EC2Client";
        var commandName = "ModifyTrafficMirrorFilterNetworkServicesCommand";
        var handlerExecutionContext = {
            logger: logger,
            clientName: clientName,
            commandName: commandName,
            inputFilterSensitiveLog: ModifyTrafficMirrorFilterNetworkServicesRequest.filterSensitiveLog,
            outputFilterSensitiveLog: ModifyTrafficMirrorFilterNetworkServicesResult.filterSensitiveLog,
        };
        var requestHandler = configuration.requestHandler;
        return stack.resolve(function (request) {
            return requestHandler.handle(request.request, options || {});
        }, handlerExecutionContext);
    };
    ModifyTrafficMirrorFilterNetworkServicesCommand.prototype.serialize = function (input, context) {
        return serializeAws_ec2ModifyTrafficMirrorFilterNetworkServicesCommand(input, context);
    };
    ModifyTrafficMirrorFilterNetworkServicesCommand.prototype.deserialize = function (output, context) {
        return deserializeAws_ec2ModifyTrafficMirrorFilterNetworkServicesCommand(output, context);
    };
    return ModifyTrafficMirrorFilterNetworkServicesCommand;
}($Command));
export { ModifyTrafficMirrorFilterNetworkServicesCommand };
